Transcription task: Allow one transcription line for each previously transcribed line

eatyourgreens opened the pull request:
This an experimental change, to suggest some possible fixes for #2082.

Give each transcription line a fixed ID, when it is created from a previous line. Observe transcription task marks for changes. Re-use an existing transcription line, rather than creating a new line, when a magenta line is clicked.

Optionally: disable or even hide magenta lines when they have a linked transcription line with the same line ID.

See #1836 for discussion of whether magenta lines should still be interactive after creating a green line.

Package:
lib-classifier

Closes #2082.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

eatyourgreens commented:
Working on this, I noticed that previous annotations are lost from marks after creating a green line. Clicking a green line opens a regular text task, with no dropdown suggestions (see 'Quirk 1: Agreement line textboxes' in https://github.com/zooniverse/front-end-monorepo/issues/1719#issuecomment-663271837)

The reason for that is that `setSubTaskVisiblity(false)` deletes any previous annotations from a line. If that's a bug, not a feature, then it could be fixed in #2080, where setting subtask visibility and setting text suggestions are split into separate actions. See https://github.com/zooniverse/front-end-monorepo/pull/2080#discussion_r601476744

srallen reviewed (changes requested):
> First: Create a green line from a magenta line, then drag the green line away from the magenta line (see #1836.) You can now click on the magenta line again, to create a second green line.

This is known and I had discussed some time ago with @snblickhan and we held off making a change regarding it since we thought there may be cases where this is desired. Since @snblickhan is on the book sprint this week and thus unavailable, could you please hold off on submitting changes related to this? 

I'm submitted a review to make sure this doesn't accidentally get merged.

srallen commented:
I've had a chance to talk with @snblickhan and also review feedback from the second beta. I've confirmed with her that we do want the new marks to be movable and linked to the original previous mark. I'm going to be out next week, so I'll be able to do an actual review when I'm back.

eatyourgreens commented:
Previous annotations are cleared from a transcription line when the subtask popup closes. Removing line 54 here would make previous annotations accessible from the green transcription lines.
https://github.com/zooniverse/front-end-monorepo/blob/ead1abcdaeb817286196a2b7921a0bbdfe2b651f/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/SubTaskPopup/SubTaskPopup.js#L49-L56

eatyourgreens commented:
If this does turn out to be useful, I've added tests and moved the new behaviour (edit an existing line vs. creating a new line) into its own component and functions.

srallen reviewed (changes requested):
I'd prefer the refactoring of the transcribed line into its own component to be reverted. I started to leave several comments about it that seems only partially completed, but this is a distraction from the focus of the PR. The focus should be linking the new mark to the previous to prevent multiple marks from being created. If you'd like to propose a refactor, I'd prefer it to be separate and to be complete with both line types of pulled into its own component and each separately connected to the store for the data each needs.

eatyourgreens commented:
No problem, I've taken those commits out. The refactor was only so I could test the new behaviour.

srallen commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/TranscribedLines/TranscribedLinesContainer.js:
```diff
@@ -38,6 +40,7 @@ function TranscribedLinesContainer (props) {
     return (
       <TranscribedLines
         lines={visibleLinesPerFrame}
+        marks={transcriptionTask.marks}
```
Non-blocking: this is a bit redundant. We're already passing the task down as a prop and marks can be destructured and used in the child component already.

eatyourgreens replied:
Yeah. Without this, I found that changes to the `marks` array aren't picked up by the observer. I'm not sure why.

srallen replied:
Oh yeah, the mobx observable doesn't deep observe: https://mobx.js.org/intercept-and-observe.html#intercept--observe-

srallen replied:
I think we get caught out on this often forgetting this part of the observer function... maybe a small comment would be helpful here to note that?

eatyourgreens replied:
Sure. It did feel a bit hacky when I wrote this, but then I wrote the whole thing very quickly just to see if it would work.

srallen replied:
I think this is why they encourage connecting the store whenever it's needed and so if this does get refactored into two smaller separate components, the connector for the transcribed line could destructor the task ahead of time and pass in everything that's needed as props rather than the task object.

eatyourgreens replied:
I've destructured the `marks` prop in the `useStores` hook and added a comment.

srallen reviewed (changes requested):
The spec file for `TranscribedLines` in the `incomplete lines` block should have an addition of a test that tests that a new mark is not created if an existing mark is present.

Another non-blocking comment: The `TranscribedLines` story for the one unfinished mark is broken, as it doesn't open up the subtask... This is broken on the currently deployed storybook however. 

I have some behavior questions that I think will have to wait for @snblickhan to try it next week, so let's get the test in at least and I think this will be ok to merge.

eatyourgreens commented:
@srallen I've added tests that count the number of marks and check that only one is created per previously-transcribed line. I also realised the popup was only being opened on `mark.finish()` so I've added a check for marks that are already finished.

srallen commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/TranscribedLines/TranscribedLines.js:
```diff
@@ -56,27 +61,32 @@ class TranscribedLines extends React.Component {
-      mark.finish()
+      if (mark.finished) {
+        const markBounds = node?.getBoundingClientRect()
+        mark.setSubTaskVisibility(true, markBounds)
```
This creates a UX where the subtask still opens and it can look like you can create a new mark again for a previous annotation that already has one. It doesn't, but it's confusing. I'd like to however get this in place and walk through these changes and get more feedback from @snblickhan this week.

eatyourgreens replied:
Yeah, I think this is very confusing too but having a line/button that's clickable yet does nothing is also confusing. If the magenta line does nothing, once a green line exists, then it should be disabled (or maybe removed completely so that there's only ever one line per annotation.)

I also find it a bit confusing that clicking a magenta line opens the popup with previous annotations attached to the mark, but clicking the green line opens the popup with the regular text editor. I guess that's so you can change the line that you previously chose? If you made a mistake, and want to choose a different line of text, you'd be expected to delete the green line and start again from the magenta line?

srallen reviewed (approved):
LGTM. There's some follow up needed for the overall UX, but this puts into place the core functionality of only allowing one new mark per previous annotation mark.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/TranscribedLines/TranscribedLines.spec.js:
```diff
@@ -233,44 +255,116 @@ describe('Component > TranscribedLines', function () {
+      })
+
+      it('should create only one mark per transcribed line', function () {
+        const eventMock = { key: 'Enter', preventDefault: sinon.spy() }
```
The key here should be a space.

eatyourgreens replied:
Also missing the event `target`.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/TranscribedLines/TranscribedLines.spec.js:
```diff
@@ -233,44 +255,116 @@ describe('Component > TranscribedLines', function () {
+      })
+
+      it('should create only one mark per transcribed line', function () {
+        const eventMock = { key: 'Enter', preventDefault: sinon.spy(), target: null }
```
This should mock a click event.
